// edge-tts-ts/dist/cli/edge-playback.js
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/cli/edge-playback.ts
var edge_playback_exports = {};
__export(edge_playback_exports, {
  createPlaybackCommand: () => createPlaybackCommand
});
module.exports = __toCommonJS(edge_playback_exports);
var import_commander = require("commander");
var fs = __toESM(require("fs"));

// src/edge-tts/communicate.ts
var import_isomorphic_ws = __toESM(require("isomorphic-ws"));

// src/edge-tts/constants.ts
var BASE_URL = "speech.platform.bing.com/consumer/speech/synthesize/readaloud";
var TRUSTED_CLIENT_TOKEN = "6A5AA1D4EAFF4E9FB37E23D68491D6F4";
var WSS_URL = `wss://${BASE_URL}/edge/v1?TrustedClientToken=${TRUSTED_CLIENT_TOKEN}`;
var VOICE_LIST = `https://${BASE_URL}/voices/list?trustedclienttoken=${TRUSTED_CLIENT_TOKEN}`;
var DEFAULT_VOICE = "en-US-EmmaMultilingualNeural";
var CHROMIUM_FULL_VERSION = "143.0.3650.75";
var CHROMIUM_MAJOR_VERSION = CHROMIUM_FULL_VERSION.split(".")[0];
var SEC_MS_GEC_VERSION = `1-${CHROMIUM_FULL_VERSION}`;
var BASE_HEADERS = {
  "User-Agent": `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/${CHROMIUM_MAJOR_VERSION}.0.0.0 Safari/537.36 Edg/${CHROMIUM_MAJOR_VERSION}.0.0.0`,
  "Accept-Encoding": "gzip, deflate, br, zstd",
  "Accept-Language": "en-US,en;q=0.9"
};
var WSS_HEADERS = {
  ...BASE_HEADERS,
  "Pragma": "no-cache",
  "Cache-Control": "no-cache",
  "Origin": "chrome-extension://jdiccldimpdaibmpdkjnbmckianbfold",
  "Sec-WebSocket-Version": "13"
};
var VOICE_HEADERS = {
  ...BASE_HEADERS,
  "Authority": "speech.platform.bing.com",
  "Sec-CH-UA": `" Not;A Brand";v="99", "Microsoft Edge";v="${CHROMIUM_MAJOR_VERSION}", "Chromium";v="${CHROMIUM_MAJOR_VERSION}"`,
  "Sec-CH-UA-Mobile": "?0",
  "Accept": "*/*",
  "Sec-Fetch-Site": "none",
  "Sec-Fetch-Mode": "cors",
  "Sec-Fetch-Dest": "empty"
};
var TICKS_PER_SECOND = 1e7;
var MP3_BITRATE_BPS = 48e3;

// src/edge-tts/drm.ts
var WIN_EPOCH = 11644473600;
var S_TO_NS = 1e9;
var _DRM = class _DRM {
  static adj_clock_skew_seconds(skew_seconds) {
    _DRM.clock_skew_seconds += skew_seconds;
  }
  static get_unix_timestamp() {
    return Date.now() / 1e3 + _DRM.clock_skew_seconds;
  }
  static parse_rfc2616_date(dateStr) {
    const timestamp = Date.parse(dateStr);
    return isNaN(timestamp) ? null : timestamp / 1e3;
  }
  static async generate_sec_ms_gec() {
    let ticks = _DRM.get_unix_timestamp();
    ticks += WIN_EPOCH;
    ticks -= ticks % 300;
    ticks *= S_TO_NS / 100;
    const strToHash = `${ticks.toFixed(0)}${TRUSTED_CLIENT_TOKEN}`;
    return await _DRM.sha256(strToHash);
  }
  static async sha256(message) {
    const msgUint8 = new TextEncoder().encode(message);
    let hashBuffer;
    if (typeof window !== "undefined" && window.crypto && window.crypto.subtle) {
      hashBuffer = await window.crypto.subtle.digest("SHA-256", msgUint8);
    } else {
      const crypto2 = await import("crypto");
      const hash = crypto2.createHash("sha256");
      hash.update(message);
      return hash.digest("hex").toUpperCase();
    }
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    const hashHex = hashArray.map((b) => b.toString(16).padStart(2, "0")).join("");
    return hashHex.toUpperCase();
  }
  static generate_muid() {
    const bytes = new Uint8Array(16);
    if (typeof window !== "undefined" && window.crypto) {
      window.crypto.getRandomValues(bytes);
    } else {
      try {
        const crypto2 = require("crypto");
        return crypto2.randomBytes(16).toString("hex").toUpperCase();
      } catch (e) {
        return Array.from(bytes).map(() => Math.floor(Math.random() * 256).toString(16).padStart(2, "0")).join("").toUpperCase();
      }
    }
    return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("").toUpperCase();
  }
  static headers_with_muid(headers) {
    const combined_headers = { ...headers };
    combined_headers["Cookie"] = `muid=${_DRM.generate_muid()};`;
    return combined_headers;
  }
};
_DRM.clock_skew_seconds = 0;
var DRM = _DRM;

// src/edge-tts/utils.ts
function removeIncompatibleCharacters(str) {
  const chars = Array.from(str);
  for (let i = 0; i < chars.length; i++) {
    const code = chars[i].charCodeAt(0);
    if (code >= 0 && code <= 8 || code >= 11 && code <= 12 || code >= 14 && code <= 31) {
      chars[i] = " ";
    }
  }
  return chars.join("");
}
function connectId() {
  const bytes = new Uint8Array(16);
  if (typeof crypto !== "undefined") {
    crypto.getRandomValues(bytes);
  } else {
    for (let i = 0; i < 16; i++) bytes[i] = Math.floor(Math.random() * 256);
  }
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function dateToString() {
  const date = /* @__PURE__ */ new Date();
  const parts = date.toUTCString().split(" ");
  const days = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
  const months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
  const d = days[date.getUTCDay()];
  const m = months[date.getUTCMonth()];
  const day = date.getUTCDate().toString().padStart(2, "0");
  const year = date.getUTCFullYear();
  const time = date.getUTCHours().toString().padStart(2, "0") + ":" + date.getUTCMinutes().toString().padStart(2, "0") + ":" + date.getUTCSeconds().toString().padStart(2, "0");
  return `${d} ${m} ${day} ${year} ${time} GMT+0000 (Coordinated Universal Time)`;
}
function escapeXml(unsafe) {
  return unsafe.replace(/[<>&'"]/g, (c) => {
    switch (c) {
      case "<":
        return "&lt;";
      case ">":
        return "&gt;";
      case "&":
        return "&amp;";
      case "'":
        return "&apos;";
      case '"':
        return "&quot;";
      default:
        return c;
    }
  });
}
function unescapeXml(safe) {
  return safe.replace(/&lt;|&gt;|&amp;|&apos;|&quot;/g, (c) => {
    switch (c) {
      case "&lt;":
        return "<";
      case "&gt;":
        return ">";
      case "&amp;":
        return "&";
      case "&apos;":
        return "'";
      case "&quot;":
        return '"';
      default:
        return c;
    }
  });
}
function* splitTextByByteLength(text, byteLength) {
  const encoder = new TextEncoder();
  const decoder = new TextDecoder();
  let data = encoder.encode(text);
  while (data.length > byteLength) {
    let splitAt = findLastNewlineOrSpaceWithinLimit(data, byteLength);
    if (splitAt < 0) {
      splitAt = findSafeUtf8SplitPoint(data, byteLength);
    }
    splitAt = adjustSplitPointForXmlEntity(data, splitAt);
    if (splitAt <= 0) {
      splitAt = 1;
    }
    const chunk = data.slice(0, splitAt);
    const chunkStr = decoder.decode(chunk).trim();
    if (chunkStr) {
      yield chunkStr;
    }
    data = data.slice(splitAt);
  }
  const remaining = decoder.decode(data).trim();
  if (remaining) {
    yield remaining;
  }
}
function findLastNewlineOrSpaceWithinLimit(data, limit) {
  const space = 32;
  const newline = 10;
  let last = -1;
  for (let i = 0; i < limit; i++) {
    if (data[i] === newline) last = i;
    else if (data[i] === space && last === -1) {
    }
  }
  for (let i = limit - 1; i >= 0; i--) {
    if (data[i] === newline) return i;
  }
  for (let i = limit - 1; i >= 0; i--) {
    if (data[i] === space) return i;
  }
  return -1;
}
function findSafeUtf8SplitPoint(data, limit) {
  let splitAt = limit;
  const decoder = new TextDecoder("utf-8", { fatal: true });
  while (splitAt > 0) {
    try {
      decoder.decode(data.slice(0, splitAt));
      return splitAt;
    } catch (e) {
      splitAt--;
    }
  }
  return splitAt;
}
function adjustSplitPointForXmlEntity(data, splitAt) {
  const amp = 38;
  const semi = 59;
  let current = splitAt;
  const sub = data.slice(0, current);
  const lastAmp = sub.lastIndexOf(amp);
  if (lastAmp !== -1) {
    const lastSemi = sub.lastIndexOf(semi);
    if (lastSemi < lastAmp) {
      return lastAmp;
    }
  }
  return current;
}

// src/edge-tts/communicate.ts
var Communicate = class {
  constructor(text, options = {}) {
    this.tts_config = {
      voice: options.voice || DEFAULT_VOICE,
      rate: options.rate || "+0%",
      volume: options.volume || "+0%",
      pitch: options.pitch || "+0Hz",
      boundary: options.boundary || "SentenceBoundary"
    };
    const cleanedText = removeIncompatibleCharacters(text);
    const escapedText = escapeXml(cleanedText);
    this.texts = Array.from(splitTextByByteLength(escapedText, 4096));
    this.state = {
      partial_text: new Uint8Array(),
      // Not really used this way in TS
      offset_compensation: 0,
      last_duration_offset: 0,
      stream_was_called: false,
      chunk_audio_bytes: 0,
      cumulative_audio_bytes: 0
    };
  }
  mkssml(escaped_text) {
    return `<speak version='1.0' xmlns='http://www.w3.org/2001/10/synthesis' xml:lang='en-US'><voice name='${this.tts_config.voice}'><prosody pitch='${this.tts_config.pitch}' rate='${this.tts_config.rate}' volume='${this.tts_config.volume}'>${escaped_text}</prosody></voice></speak>`;
  }
  ssml_headers_plus_data(request_id, timestamp, ssml) {
    return `X-RequestId:${request_id}\r
Content-Type:application/ssml+xml\r
X-Timestamp:${timestamp}Z\r
Path:ssml\r
\r
${ssml}`;
  }
  parseMetadata(data) {
    const metadata = JSON.parse(data);
    for (const meta_obj of metadata.Metadata) {
      const meta_type = meta_obj.Type;
      if (meta_type === "WordBoundary" || meta_type === "SentenceBoundary") {
        const current_offset = meta_obj.Data.Offset + this.state.offset_compensation;
        const current_duration = meta_obj.Data.Duration;
        return {
          type: meta_type,
          offset: current_offset,
          duration: current_duration,
          text: unescapeXml(meta_obj.Data.text.Text)
        };
      }
    }
    throw new Error("No valid metadata found");
  }
  compensateOffset() {
    this.state.cumulative_audio_bytes += this.state.chunk_audio_bytes;
    this.state.offset_compensation = Math.floor(
      this.state.cumulative_audio_bytes * 8 * TICKS_PER_SECOND / MP3_BITRATE_BPS
    );
    this.state.chunk_audio_bytes = 0;
  }
  async *stream() {
    if (this.state.stream_was_called) {
      throw new Error("stream can only be called once.");
    }
    this.state.stream_was_called = true;
    for (const textChunk of this.texts) {
      this.state.chunk_audio_bytes = 0;
      yield* this._stream(textChunk);
    }
  }
  async *_stream(textChunk) {
    const requestId = connectId();
    const url = `${WSS_URL}&ConnectionId=${requestId}&Sec-MS-GEC=${await DRM.generate_sec_ms_gec()}&Sec-MS-GEC-Version=${SEC_MS_GEC_VERSION}`;
    const headers = DRM.headers_with_muid(WSS_HEADERS);
    const ws = new import_isomorphic_ws.default(url, {
      headers
    });
    ws.binaryType = "arraybuffer";
    let resolveOpen;
    let rejectOpen;
    const openPromise = new Promise((res, rej) => {
      resolveOpen = res;
      rejectOpen = rej;
    });
    ws.onopen = () => resolveOpen();
    ws.onerror = (err) => rejectOpen(err);
    await openPromise;
    const sendCommand = () => {
      const word_boundary = this.tts_config.boundary === "WordBoundary";
      const wd = word_boundary ? "true" : "false";
      const sq = !word_boundary ? "true" : "false";
      ws.send(
        `X-Timestamp:${dateToString()}\r
Content-Type:application/json; charset=utf-8\r
Path:speech.config\r
\r
{"context":{"synthesis":{"audio":{"metadataoptions":{"sentenceBoundaryEnabled":"${sq}","wordBoundaryEnabled":"${wd}"},"outputFormat":"audio-24khz-48kbitrate-mono-mp3"}}}}`
      );
    };
    const sendSSML = () => {
      ws.send(
        this.ssml_headers_plus_data(
          connectId(),
          dateToString(),
          this.mkssml(textChunk)
        )
      );
    };
    sendCommand();
    sendSSML();
    const queue = [];
    let resolveNext = null;
    let error = null;
    ws.onmessage = (event) => {
      if (typeof event.data === "string") {
        const headEnd = event.data.indexOf("\r\n\r\n");
        const headersStr = event.data.substring(0, headEnd);
        const body = event.data.substring(headEnd + 4);
        const headers2 = {};
        headersStr.split("\r\n").forEach((line) => {
          const [k, v] = line.split(":", 2);
          if (k) headers2[k.trim()] = v ? v.trim() : "";
        });
        const path = headers2["Path"];
        if (path === "audio.metadata") {
          queue.push(this.parseMetadata(body));
        } else if (path === "turn.end") {
          this.compensateOffset();
          queue.push("DONE");
        }
      } else if (event.data instanceof ArrayBuffer) {
        const view = new DataView(event.data);
        const headerLength = view.getUint16(0);
        const headersData = new TextDecoder().decode(event.data.slice(2, 2 + headerLength));
        const audioData = new Uint8Array(event.data.slice(2 + headerLength));
        const headers2 = {};
        headersData.split("\r\n").forEach((line) => {
          const [k, v] = line.split(":", 2);
          if (k) headers2[k.trim()] = v ? v.trim() : "";
        });
        if (headers2["Path"] === "audio") {
          if (audioData.length > 0) {
            this.state.chunk_audio_bytes += audioData.length;
            queue.push({ type: "audio", data: audioData });
          }
        }
      }
      if (resolveNext) resolveNext();
    };
    ws.onerror = (err) => {
      error = err;
      queue.push("ERROR");
      if (resolveNext) resolveNext();
    };
    ws.onclose = () => {
      if (!queue.includes("DONE") && !queue.includes("ERROR")) {
        queue.push("DONE");
      }
      if (resolveNext) resolveNext();
    };
    while (true) {
      if (queue.length === 0) {
        await new Promise((res) => {
          resolveNext = res;
        });
        resolveNext = null;
      }
      const item = queue.shift();
      if (item === "DONE") break;
      if (item === "ERROR") throw error || new Error("WebSocket error");
      if (item) yield item;
    }
    ws.close();
  }
};

// src/edge-playback/util.ts
function prErr(msg) {
  console.error(`error: ${msg}`);
}

// src/edge-playback/play.ts
async function playAudio(audioData) {
  if (typeof window !== "undefined" && typeof Blob !== "undefined") {
    const blob = new Blob([audioData], { type: "audio/mpeg" });
    const url = URL.createObjectURL(blob);
    const audio = new Audio(url);
    return new Promise((resolve, reject) => {
      audio.onended = () => {
        URL.revokeObjectURL(url);
        resolve();
      };
      audio.onerror = (e) => {
        prErr("Audio playback error");
        reject(e);
      };
      audio.src = url;
      audio.play().then(resolve).catch(reject);
    });
  } else {
    const fs2 = await import("fs");
    const path = await import("path");
    const os = await import("os");
    const soundPlay = await import("sound-play");
    const tempDir = os.tmpdir();
    const tempFile = path.join(tempDir, `edge_tts_${Date.now()}.mp3`);
    fs2.writeFileSync(tempFile, Buffer.from(audioData));
    try {
      await soundPlay.default.play(tempFile);
    } catch (err) {
      prErr(`Failed to play audio: ${err}`);
      throw err;
    } finally {
      try {
        fs2.unlinkSync(tempFile);
      } catch (e) {
      }
    }
  }
}

// src/cli/edge-playback.ts
function createPlaybackCommand() {
  const program = new import_commander.Command();
  program.name("edge-playback").description("Microsoft Edge TTS Playback CLI").option("-t, --text <text>", "Text to synthesize").option("-f, --file <file>", "File to read text from").option("-v, --voice <voice>", "Voice to use", "en-US-EmmaMultilingualNeural").option("--rate <rate>", "Rate of speech", "+0%").option("--volume <volume>", "Volume of speech", "+0%").option("--pitch <pitch>", "Pitch of speech", "+0Hz").action(async (options) => {
    let text = options.text;
    if (options.file) {
      text = fs.readFileSync(options.file, "utf-8");
    }
    if (!text) {
      console.error("Text or file is required");
      process.exit(1);
    }
    const comm = new Communicate(text, {
      voice: options.voice,
      rate: options.rate,
      volume: options.volume,
      pitch: options.pitch
    });
    const audioChunks = [];
    for await (const chunk of comm.stream()) {
      if (chunk.type === "audio") {
        audioChunks.push(chunk.data);
      }
    }
    const totalLength = audioChunks.reduce((acc, c) => acc + c.length, 0);
    const audioData = new Uint8Array(totalLength);
    let offset = 0;
    for (const chunk of audioChunks) {
      audioData.set(chunk, offset);
      offset += chunk.length;
    }
    try {
      console.log("Playing audio...");
      await playAudio(audioData);
    } catch (err) {
    }
  });
  return program;
}
if (require.main === module) {
  createPlaybackCommand().parse(process.argv);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createPlaybackCommand
});
