// edge-tts-ts/dist/chunk-ZJDB53YB.mjs
import {
  DRM,
  SEC_MS_GEC_VERSION,
  VOICE_HEADERS,
  VOICE_LIST
} from "./chunk-FOKMJMUX.mjs";

// src/edge-tts/voices.ts
async function listVoices(options = {}) {
  const url = new URL(VOICE_LIST);
  url.searchParams.set("Sec-MS-GEC", await DRM.generate_sec_ms_gec());
  url.searchParams.set("Sec-MS-GEC-Version", SEC_MS_GEC_VERSION);
  const headers = DRM.headers_with_muid(VOICE_HEADERS);
  const response = await fetch(url.toString(), {
    headers
  });
  if (!response.ok) {
    if (response.status === 403) {
      const dateHeader = response.headers.get("Date");
      if (dateHeader) {
        const serverTime = DRM.parse_rfc2616_date(dateHeader);
        if (serverTime) {
          DRM.adj_clock_skew_seconds(serverTime - DRM.get_unix_timestamp());
          return listVoices(options);
        }
      }
    }
    throw new Error(`Failed to list voices: ${response.statusText}`);
  }
  const data = await response.json();
  for (const voice of data) {
    if (!voice.VoiceTag) {
      voice.VoiceTag = { ContentCategories: [], VoicePersonalities: [] };
    }
  }
  return data;
}
var VoicesManager = class _VoicesManager {
  constructor() {
    this.voices = [];
    this.calledCreate = false;
  }
  static async create(customVoices) {
    const manager = new _VoicesManager();
    const voices = customVoices || await listVoices();
    manager.voices = voices.map((voice) => ({
      ...voice,
      Language: voice.Locale.split("-")[0]
    }));
    manager.calledCreate = true;
    return manager;
  }
  find(criteria) {
    if (!this.calledCreate) {
      throw new Error("VoicesManager.find() called before VoicesManager.create()");
    }
    return this.voices.filter((voice) => {
      for (const [key, value] of Object.entries(criteria)) {
        if (voice[key] !== value) {
          return false;
        }
      }
      return true;
    });
  }
};

// src/edge-tts/srt-composer.ts
function timedeltaToSrtTimestamp(seconds) {
  const hrs = Math.floor(seconds / 3600);
  const mins = Math.floor(seconds % 3600 / 60);
  const secs = Math.floor(seconds % 60);
  const msecs = Math.floor(seconds % 1 * 1e3);
  return `${hrs.toString().padStart(2, "0")}:${mins.toString().padStart(2, "0")}:${secs.toString().padStart(2, "0")},${msecs.toString().padStart(3, "0")}`;
}
function makeLegalContent(content) {
  if (content && content[0] !== "\n" && !content.includes("\n\n")) {
    return content;
  }
  return content.replace(/\n\n+/g, "\n").trim();
}
function subtitleToSrt(sub, eol = "\n") {
  const content = makeLegalContent(sub.content).replace(/\n/g, eol);
  return `${sub.index || 0}${eol}${timedeltaToSrtTimestamp(sub.start)} --> ${timedeltaToSrtTimestamp(sub.end)}${eol}${content}${eol}${eol}`;
}
function compose(subtitles, options = {}) {
  let subs = [...subtitles];
  if (options.reindex !== false) {
    subs.sort((a, b) => a.start - b.start || a.end - b.end);
    const startIndex = options.startIndex ?? 1;
    subs = subs.map((sub, i) => ({ ...sub, index: startIndex + i }));
  }
  const eol = options.eol || "\n";
  return subs.map((sub) => subtitleToSrt(sub, eol)).join("");
}

// src/edge-tts/submaker.ts
var SubMaker = class {
  constructor() {
    this.cues = [];
    this.type = null;
  }
  feed(msg) {
    if (msg.type !== "WordBoundary" && msg.type !== "SentenceBoundary") {
      throw new Error("Invalid message type, expected 'WordBoundary' or 'SentenceBoundary'.");
    }
    if (this.type === null) {
      this.type = msg.type;
    } else if (this.type !== msg.type) {
      throw new Error(`Expected message type '${this.type}', but got '${msg.type}'.`);
    }
    this.cues.push({
      index: this.cues.length + 1,
      start: msg.offset / 1e7,
      end: (msg.offset + msg.duration) / 1e7,
      content: msg.text
    });
  }
  getSrt() {
    return compose(this.cues);
  }
  toString() {
    return this.getSrt();
  }
};

export {
  listVoices,
  VoicesManager,
  timedeltaToSrtTimestamp,
  makeLegalContent,
  subtitleToSrt,
  compose,
  SubMaker
};
