// edge-tts-ts/dist/index.mjs
import {
  SubMaker,
  VoicesManager,
  compose,
  listVoices,
  makeLegalContent,
  subtitleToSrt,
  timedeltaToSrtTimestamp
} from "./chunk-ZJDB53YB.mjs";
import {
  playAudio,
  prErr
} from "./chunk-6PHLBSK4.mjs";
import {
  BASE_HEADERS,
  BASE_URL,
  CHROMIUM_FULL_VERSION,
  CHROMIUM_MAJOR_VERSION,
  Communicate,
  DEFAULT_VOICE,
  DRM,
  MP3_BITRATE_BPS,
  SEC_MS_GEC_VERSION,
  S_TO_NS,
  TICKS_PER_SECOND,
  TRUSTED_CLIENT_TOKEN,
  VOICE_HEADERS,
  VOICE_LIST,
  WIN_EPOCH,
  WSS_HEADERS,
  WSS_URL,
  connectId,
  dateToString,
  escapeXml,
  removeIncompatibleCharacters,
  splitTextByByteLength,
  unescapeXml
} from "./chunk-FOKMJMUX.mjs";
export {
  BASE_HEADERS,
  BASE_URL,
  CHROMIUM_FULL_VERSION,
  CHROMIUM_MAJOR_VERSION,
  Communicate,
  DEFAULT_VOICE,
  DRM,
  MP3_BITRATE_BPS,
  SEC_MS_GEC_VERSION,
  S_TO_NS,
  SubMaker,
  TICKS_PER_SECOND,
  TRUSTED_CLIENT_TOKEN,
  VOICE_HEADERS,
  VOICE_LIST,
  VoicesManager,
  WIN_EPOCH,
  WSS_HEADERS,
  WSS_URL,
  compose,
  connectId,
  dateToString,
  escapeXml,
  listVoices,
  makeLegalContent,
  playAudio,
  prErr,
  removeIncompatibleCharacters,
  splitTextByByteLength,
  subtitleToSrt,
  timedeltaToSrtTimestamp,
  unescapeXml
};
